design-proposals/vdpa-interfaces.md

```diff
@@ -0,0 +1,146 @@
+
+## Non Goals
+Because the implementation of vDPA can be SR-IOV based or not, this mechanism should not be coupled with the [Intel SR-IOV device plugin](https://github.com/intel/sriov-network-device-plugin) and [SR-IOV CNI plugin](https://github.com/k8snetworkplumbingwg/sriov-cni).
+
+## User Stories
+As a Kubevirt user I would like to create a vm with vdpa interfaces.
+
+## Repos
+[KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+Design can be inspired by the existing [sriov type interface](https://kubevirt.io/user-guide/virtual_machines/interfaces_and_networks/#sriov), but you can see the difference by comparing the libvirt domain XML:
+```xml
+<hostdev type="pci" managed="no" mode="subsystem">
+        <source>
+                <address type="" domain="0x0000" bus="0xcc" slot="0x00" function="0x4"></address>
+        </source>
+        <address type="pci" domain="0x0000" bus="0x07" slot="0x00" function="0x0"></address>
+        <alias name="ua-sriov-offload-ovn"></alias>
+</hostdev>
+```
+```xml
+<interface type="vdpa">
+        <source dev="/dev/vhost-vdpa-1"></source>
+        <model type="virtio"></model>
+        <alias name="ua-offload-ovn"></alias>
+        <driver name="vhost" queues="4"></driver>
+</interface>
+```
+- The former is hostdev while the latter is interface
+- The source of the former is a bdf while the source of the latter is a device path.
+
+## API
+Add a vDPA type to InterfaceBindingMethod:
+```go
+// Represents the method which will be used to connect the interface to the guest.
+// Only one of its members may be specified.
+type InterfaceBindingMethod struct {
+	Bridge     *InterfaceBridge     `json:"bridge,omitempty"`
+	Slirp      *InterfaceSlirp      `json:"slirp,omitempty"`
+	Masquerade *InterfaceMasquerade `json:"masquerade,omitempty"`
+	SRIOV      *InterfaceSRIOV      `json:"sriov,omitempty"`
+	VDPA       *InterfaceVDPA       `json:"vdpa,omitempty"`
+	Macvtap    *InterfaceMacvtap    `json:"macvtap,omitempty"`
+	Passt      *InterfacePasst      `json:"passt,omitempty"`
+}
+```
+
+## Example
+VMI spec with two vDPA devices, connected to the same network:
+```yaml
+          networkInterfaceMultiqueue: true
+          interfaces:
+          - masquerade: {}
+            name: default
+          - name: offload
+            vdpa: {}
+          - name: offload2
+            vdpa: {}
+      networks:
+      - name: default
+        pod: {}
+      - multus:
+          networkName: default/offload-ovn
+        name: offload
+      - multus:
+          networkName: default/offload-ovn
+        name: offload2
+```
+
+The default/offload-ovn NetworkAttachmentDefinition spec:
+```yaml
+apiVersion: k8s.cni.cncf.io/v1
+kind: NetworkAttachmentDefinition
+metadata:
+  name: offload-ovn
+  namespace: default
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: intel.com/vdpa_jaguar_vhost
+spec:
+  config: >-
+    {
+        "cniVersion": "0.3.0",
+        "type": "kube-ovn",
```

Comment:
typo ?  u wanna use  "sriov" here  ?

Reply:
@zzfancy it's not a typo. sriov-cni also can be used with [this](https://github.com/redhat-nfvpe/vdpa-deployment#sriov-cni). kube-ovn can offload ovs to our smart nic.